sysfs/rapl.go

```diff
@@ -0,0 +1,139 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !windows
+
+package sysfs
+
+import (
+	"errors"
+	"io/ioutil"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/procfs/internal/util"
+)
+
+// RaplValue stores the information for one RAPL power counter value
+type RaplValue struct {
+	Name           string // name of RAPL value from file "name"
+	Index          int    // index (different value for duplicate names)
+	Microjoules    uint64 // RAPL microjoule value
+	MaxMicrojoules uint64 // max RAPL microjoule value
+}
+
+// internal raplValue which contains also the sysfs filename
+type raplValue struct {
+	sysfsFilename string
+	RaplValue
+}
+
+// RaplReader reads RaplValues from sysfs
+type RaplReader struct {
+	raplValues []raplValue
+}
+
+// NewRaplReader returns a new reader for RAPL values.
+// When RAPL files are not present, returns nil with error
+// https://www.kernel.org/doc/Documentation/power/powercap/powercap.txt
+func NewRaplReader(fs FS) (*RaplReader, error) {
+	reader := RaplReader{}
+	raplDir := fs.sys.Path("class/powercap")
+
+	files, err := ioutil.ReadDir(raplDir)
+	if err != nil {
+		return nil, errors.New(
+			"no sysfs powercap / RAPL power metrics files found")
+	}
+
+	// count name usages to avoid duplicates (label them with an index)
+	countNameUsages := make(map[string]int)
+
+	// loop through directory files searching for file "name" from subdirs
+	for _, f := range files {
+		nameFile := filepath.Join(raplDir, f.Name(), "/name")
+		nameBytes, err := ioutil.ReadFile(nameFile)
+		if err == nil {
+			// add new rapl value since name file was found
+			name := strings.TrimSpace(string(nameBytes))
+
+			// make RAPL name prometheus-compatible
+			modifiedPromName := strings.Replace(name, "-", "_", -1)
+
+			// get a pair of index and final name
+			index, modifiedPromName := getIndexAndName(countNameUsages,
+				modifiedPromName)
+
+			maxMicrojouleFilename := filepath.Join(raplDir, f.Name(),
+				"/max_energy_range_uj")
+			microjouleFilename := filepath.Join(raplDir, f.Name(),
+				"/energy_uj")
+			maxMicrojoules, err := util.ReadUintFromFile(maxMicrojouleFilename)
+			if err != nil {
+				return nil, err
+			}
+
+			rv := RaplValue{
+				Name:           modifiedPromName,
+				Index:          index,
+				Microjoules:    0,
+				MaxMicrojoules: maxMicrojoules,
+			}
+			privRv := raplValue{
+				sysfsFilename: microjouleFilename,
+				RaplValue:     rv,
+			}
+
+			reader.raplValues = append(reader.raplValues, privRv)
+
+			// Store into map how many times this name has been used. There can
+			// be e.g. multiple "dram" instances without any index postfix. The
+			// count is then used for indexing
+			countNameUsages[modifiedPromName] = index + 1
+		}
+	}
+
+	return &reader, nil
```

Comment:
Why do you return a reader here instead of just reading and setting the Microjoules value right away?

Reply:
It is more efficient to resolve the duplicate name issue and possible name postfix parsing just once, instead of every time the values are read. This is what I intended for the API usage:

1) user creates a RaplReader, which at that point goes through the files and figures out which names are duplicates and which need their postfix (_0, _1) removed and delivered as an index instead (that value can be used as a label then).
2) user calls NewRaplValues which only goes through the slice of known values for which all duplicate-issues and indexing has been resolved earlier. The function only reads the values from the known paths, as fast as possible.

Now, the added API-complexity may not be justified by the amount of saved CPU-cycles. If you want, I can combine the functions. Your call.

Reply:
Ok, makes sense, you are avoiding re-reading some of the static information such as the directory structure and max energy since those doesn't normally change between reads.  Maybe some slightly different naming of can make this more clear.  What do you think about instead of having `RaplValue` and `raplValue` structs, there could be a `RaplZone` struct which contains the static info such as the name, max energy, and the path?  You would have a function like `GetRaplZones` which returns a slice of the RaplZones, then you can have a method of RaplZone which reads the current state.

```go
func (z *RaplZone) GetEnergyMicroJoules() (uint64, err) {
    return util.ReadUintFromFile(filepath.Join(z.Path, "energy_uj")
}
```
In this case the caller would be responsible for looping through the zones, which I think would be ok, since they would need to loop through the results in either case. WDYT?

Reply:
That sounds fine by me. I was attempting to make an effort at hiding any procfs-details from the API user, such as the paths, but having the path in the struct does make it simpler and your proposed API is very clearly named. I'll upload a version with these changes next.

Reply:
See 
2192ec2